pkg/zabbix_api.go

```diff
@@ -250,3 +251,192 @@ func makeHttpRequest(ctx context.Context, req *http.Request) ([]byte, error) {
+	} else if len(appids) > 0 {
+		allItems, err = ds.getAllItems(ctx, dsInfo, nil, appids, itemType)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	var items []*simplejson.Json
+	for k := range allItems.Get("result").MustArray() {
+		if allItems.Get("result").Get("status").MustString() == "0" {
+			matched, err := regexp.MatchString(itemFilter, allItems.Get("result").GetIndex(k).MustString())
+			if err != nil {
+				return nil, err
+			} else if matched {
+				items = append(items, allItems.Get("result").GetIndex(k))
+			}
+		}
+	}
+	return items, nil
+}
+
+func (ds *ZabbixDatasource) getApps(ctx context.Context, dsInfo *datasource.DatasourceInfo, groupFilter string, hostFilter string, appFilter string) ([]*simplejson.Json, error) {
+	hosts, err := ds.getHosts(ctx, dsInfo, groupFilter, hostFilter)
+	if err != nil {
+		return nil, err
+	}
+	var hostids []string
+	for i := range hosts {
+		hostids = append(hostids, hosts[i].Get("hostid").MustString())
+	}
+	allApps, err := ds.getAllApps(ctx, dsInfo, hostids)
+	if err != nil {
+		return nil, err
+	}
+	var apps []*simplejson.Json
+	for k := range allApps.Get("result").MustArray() {
+		matched, err := regexp.MatchString(appFilter, allApps.Get("result").GetIndex(k).MustString())
+		if err != nil {
+			return nil, err
+		} else if matched {
+			apps = append(apps, allApps.Get("result").GetIndex(k))
+		}
+	}
+	return apps, nil
+}
+
+func (ds *ZabbixDatasource) getHosts(ctx context.Context, dsInfo *datasource.DatasourceInfo, groupFilter string, hostFilter string) ([]*simplejson.Json, error) {
+	groups, err := ds.getGroups(ctx, dsInfo, groupFilter)
+	if err != nil {
+		return nil, err
+	}
+	var groupids []string
+	for i := range groups {
+		groupids = append(groupids, groups[i].Get("groupid").MustString())
+	}
+	allHosts, err := ds.getAllHosts(ctx, dsInfo, groupids)
+	if err != nil {
+		return nil, err
+	}
+	var hosts []*simplejson.Json
+	for k := range allHosts.Get("result").MustArray() {
+		matched, err := regexp.MatchString(hostFilter, allHosts.Get("result").GetIndex(k).MustString())
+		if err != nil {
+			return nil, err
+		} else if matched {
+			hosts = append(hosts, allHosts.Get("result").GetIndex(k))
+		}
+	}
+	return hosts, nil
+}
+
+func (ds *ZabbixDatasource) getGroups(ctx context.Context, dsInfo *datasource.DatasourceInfo, groupFilter string) ([]*simplejson.Json, error) {
+	allGroups, err := ds.getAllGroups(ctx, dsInfo)
+	if err != nil {
+		return nil, err
+	}
+	var groups []*simplejson.Json
+	for k := range allGroups.Get("result").MustArray() {
+		matched, err := regexp.MatchString(groupFilter, allGroups.Get("result").GetIndex(k).MustString())
+		if err != nil {
+			return nil, err
+		} else if matched {
+			groups = append(groups, allGroups.Get("result").GetIndex(k))
+		}
+	}
+	return groups, nil
+}
+
+func (ds *ZabbixDatasource) getAllItems(ctx context.Context, dsInfo *datasource.DatasourceInfo, hostids []string, appids []string, itemtype string) (*simplejson.Json, error) {
```

Comment:
Is "getAllItems" a good name for this? It's not really getting *all* items, usually it's filtered by hostids.